File: benchmarks/micro/AbstractionGenericWrapper.c
#include <stdint.h>
#include <unistd.h>

static int64_t identity_i64(int64_t value) {
    return value;
}

static int64_t mix_core(int64_t value, int64_t salt) {
    return (((value * 31) + salt) ^ (value >> 3)) % 1000003;
}

static int64_t mix(int64_t value, int64_t salt, int32_t tag) {
    (void)tag;
    return mix_core(identity_i64(value), identity_i64(salt));
}

int main(void) {
    int64_t total = (int64_t)getpid();

    for (int32_t i = 0; i < 200000; i += 1) {
        total += mix((int64_t)i, total % 97, i);
        total %= 1000000007;
    }

    return total <= 0 ? 1 : 0;
}

File: benchmarks/micro/AbstractionGenericWrapper.rs
fn identity<T>(value: T) -> T {
    value
}

fn mix_core(value: i64, salt: i64) -> i64 {
    (((value * 31) + salt) ^ (value >> 3)) % 1_000_003
}

fn mix<T>(value: i64, salt: i64, _tag: T) -> i64 {
    mix_core(identity(value), identity(salt))
}

fn main() {
    let mut total = std::process::id() as i64;

    for i in 0_i32..200_000 {
        total += mix(i as i64, total % 97, i);
        total %= 1_000_000_007;
    }

    if total <= 0 {
        std::process::exit(1);
    }
}
